src/fluree/db/json_ld/vocab.cljc

```diff
@@ -243,29 +243,62 @@
+(defn predicate-flakes-by-type
+  "Returns a map of predicate flakes mapped by type"
   [flakes]
-  (into #{}
-        (comp (filter flake/op)
-              (map
-                (fn [f]
-                  (let [p (flake/p f)
-                        o (flake/o f)]
-                    (->> [p
-                          ;; if p is a predicate ref, we know o is a predicate sid as well
-                          (when (contains? jld-ledger/predicate-refs p)
-                            o)
-                          ;; if p (type) has an o that says s is a predicate, include s as well
-                          (when (and (= p const/$rdf:type)
+  (reduce (fn [ps f]
+            (if (flake/op f)
+              (let [p         (flake/p f)
+                    o         (flake/o f)
+                    _         (log/trace "predicate-flakes-by-type p:" p "o:" o)
+                    pred-type (cond
+                                (and (= const/$rdf:type p)
                                      (jld-ledger/class-or-property-sid o))
-                            (flake/s f))]
-                         (remove nil?)))))
-              cat)
-        flakes))
+                                (do
+                                  (log/trace "predicate-flakes-by-type found ::class-or-prop pred")
+                                  ::class-or-prop)
+
+                                (jld-ledger/predicate-refs p)
+                                ::ref)]
+                (if pred-type
+                  (update ps pred-type (fnil conj #{}) f)
+                  ps))
+              ps))
+          {} flakes))
+
+(defn filter-vocab-flakes
+  "Filters out the vocab-flakes from all-flakes using the predicate-flakes map
+  (as returned by predicate-flakes-by-type)."
+  [predicate-flakes all-flakes]
+  (let [ref-flakes               (::ref predicate-flakes)
+        class-or-prop-flakes     (::class-or-prop predicate-flakes)
+        ref-flake-os             (->> ref-flakes (map flake/o) set)
+        class-or-prop-flake-sids (->> class-or-prop-flakes (map flake/s) set)
+        predicate-sids           (->> all-flakes (map flake/p) set)
+        set-conj                 (fnil conj #{})]
+    (reduce (fn [vfs f]
+              (let [fs (flake/s f)]
+                (cond
+                  (predicate-sids fs)
+                  (do
+                    (log/trace "filter-vocab-flakes adding" f "b/c its sid matches a pid")
+                    (set-conj vfs f))
+
+                  (ref-flake-os fs)
+                  (do
+                    (log/trace "filter-vocab-flakes adding" f "b/c its sid matches a ref flake's o")
+                    (set-conj vfs f))
+
+                  (class-or-prop-flake-sids fs)
+                  (do
+                    (log/trace "filter-vocab-flakes adding" f "b/c its sid matches a type flake's sid")
+                    (set-conj vfs f))
+
+                  :else vfs)))
+            ref-flakes all-flakes)))
```

Comment:
If we do a `(or ref-flakes #{})` here we can save the overhead of a `nil` check on every `conj`.